x/ccv/provider/keeper/double_vote.go

```diff
@@ -0,0 +1,110 @@
+package keeper
+
+import (
+	"bytes"
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	tmtypes "github.com/cometbft/cometbft/types"
+
+	"github.com/cosmos/interchain-security/v3/x/ccv/provider/types"
+	ccvtypes "github.com/cosmos/interchain-security/v3/x/ccv/types"
+)
+
+// HandleConsumerDoubleVoting verifies a double voting evidence for a given a consumer chain ID
+// and a public key and, if successful, executes the jailing of the malicious validator.
+func (k Keeper) HandleConsumerDoubleVoting(
+	ctx sdk.Context,
+	evidence *tmtypes.DuplicateVoteEvidence,
+	chainID string,
+	pubkey cryptotypes.PubKey,
+) error {
+	// verifies the double voting evidence using the consumer chain public key
+	if err := k.VerifyDoubleVotingEvidence(ctx, *evidence, chainID, pubkey); err != nil {
+		return err
+	}
+
+	// get the validator's consensus address on the provider
+	providerAddr := k.GetProviderAddrFromConsumerAddr(
+		ctx,
+		chainID,
+		types.NewConsumerConsAddress(sdk.ConsAddress(evidence.VoteA.ValidatorAddress.Bytes())),
+	)
+
+	if err := k.SlashValidator(ctx, providerAddr); err != nil {
+		return err
+	}
+	if err := k.JailAndTombstoneValidator(ctx, providerAddr); err != nil {
+		return err
+	}
+
+	k.Logger(ctx).Info(
+		"confirmed equivocation",
+		"byzantine validator address", providerAddr.String(),
+	)
+
+	return nil
+}
+
+// VerifyDoubleVotingEvidence verifies a double voting evidence
+// for a given chain id and a validator public key
+func (k Keeper) VerifyDoubleVotingEvidence(
+	ctx sdk.Context,
+	evidence tmtypes.DuplicateVoteEvidence,
+	chainID string,
+	pubkey cryptotypes.PubKey,
+) error {
+	if pubkey == nil {
+		return fmt.Errorf("validator public key cannot be empty")
+	}
+
+	// Note that since we're only jailing validators for double voting on a consumer chain,
+	// the age of the evidence is irrelevant and therefore isn't checked.
+
+	// H/R/S must be the same
```

Comment:
Why do we use `S` here in `H/R/S`. I guess `H` stands for `Height`, `R` for `Round`, but `S` doesn't correspond to `Type`?

Reply:
Comment taken from [CometBFT](https://github.com/cometbft/cometbft/blob/v0.37.2/evidence/verify.go#L169), I guess `S` stands for `SignMsgType`.

Reply:
I see. It still might make sense to just have `Height/Round/Type` here  even if CometBFT has it otherwise.